### PR TITLE
Add tests for List Users Admin API

### PR DESCRIPTION
**Pull request by dklimpel**
Add tests to `test_user.UsersListTestCase` for List Users Admin API
Also fix a small bug with `next_token`. `next_token` also appears at the end of the list.

This is a preparation to add a parameter  `order_by`.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: tests/rest/admin/test_user.py [hunk 1]
```diff
@@ -560,32 +551,221 @@ def _search_test(
...
+        channel = self.make_request(
+            "GET", self.url + "?guests=not_bool", access_token=self.admin_user_tok,
+        )
+
+        self.assertEqual(400, int(channel.result["code"]), msg=channel.result["body"])
+        self.assertEqual(Codes.UNKNOWN, channel.json_body["errcode"])
+        self.assertEqual(
+            "Boolean query parameter b'guests' must be one of ['true', 'false']",
+            channel.json_body["error"],
+        )
```
Comment (clokep) [inline tests/rest/admin/test_user.py:616]: Generally I don't love testing against errors like this, it seems that parameter returned should be `Codes.INVALID_PARAM` for these two boolean cases though. (Also the `b'guests'` seems like an unrelated bug -- that should probably be `'guests'`).

File: tests/rest/admin/test_user.py [hunk 2]
```diff
@@ -560,32 +551,221 @@ def _search_test(
...
+
+    def test_limit(self):
+        """
+        Testing list of users with limit
+        """
+
+        number_users = 20
+        # reduction of one admin user
```
Comment (clokep) [inline tests/rest/admin/test_user.py:636]:
```suggestion
        # Create one less user (since there's already an admin user).
```

Please update this comment in each spot.

File: tests/rest/admin/test_user.py [hunk 3]
```diff
@@ -560,32 +551,221 @@ def _search_test(
...
+        )
+
+        self.assertEqual(200, int(channel.result["code"]), msg=channel.result["body"])
+        self.assertEqual(channel.json_body["total"], number_users)
+        self.assertEqual(len(channel.json_body["users"]), 19)
+        self.assertEqual(channel.json_body["next_token"], "19")
+
+        # Check
+        # Set `from` to value of `next_token` for request remaining entries
+        #  `next_token` does not appear
```
Comment (clokep) [inline tests/rest/admin/test_user.py:731]: This seems to repeat `test_from`, but I don't see one where there's limit + from set?

**Review by clokep — CHANGES_REQUESTED** [review]
Looks reasonable overall! Thanks so much for adding tests! I left a few minor comments.

File: tests/rest/admin/test_user.py (hunk 1 above)
Comment (dklimpel) [reply tests/rest/admin/test_user.py:616]: The error is thrown here:
https://github.com/matrix-org/synapse/blob/42d3a28d8bb8c08e9e0d00a2e247cbbddb1a155c/synapse/http/servlet.py#L88-L106

I am remove the test against the string.

Comment (clokep) [reply tests/rest/admin/test_user.py:616]: Oh hrm, that would probably affect quite a few endpoints by modifying that. I suspect it is OK to make it more specific though.

File: tests/rest/admin/test_user.py [hunk 4]
```diff
@@ -560,32 +551,221 @@ def _search_test(
...
+        for u in content:
+            self.assertIn("name", u)
+            self.assertIn("is_guest", u)
+            self.assertIn("admin", u)
+            self.assertIn("user_type", u)
+            self.assertIn("deactivated", u)
+            self.assertIn("displayname", u)
+            self.assertIn("avatar_url", u)
+
+    def _create_users(self, number_users: int):
```
[the remark this answers is not in the document]
Comment (dklimpel) [reply tests/rest/admin/test_user.py:755]: I had think about it and have decided to omit it. The IDs is only needed in this one test `test_search_term`. The other tests do not need the IDs. The most work this function does is for the creation of users for the test of pagination. If you like, I can split it in two functions.  First create a user and returns ID. The second function call the first one, creates more than one user and does not return the ID.
Or do you have an other suggestion?

File: tests/rest/admin/test_user.py (hunk 3 above)
Comment (dklimpel) [reply tests/rest/admin/test_user.py:731]: This test checks that the second page shows the missing users. It checks `from`

We have 20 users. The first page (test before) request 19 users. `next_token` is 19. Now is the new requestet with `from` 19 and 1 missing user is displayed. `limit` ist not needed. Default is greater than 1 left user.

File: tests/rest/admin/test_user.py (hunk 1 above)
Comment (dklimpel) [reply tests/rest/admin/test_user.py]: I have removed it. A test on strings is really not nice. Then there are 2 less faulty tests that need to be updated when you adjust the error.

**Comment by clokep** [comment]
@dklimpel Is this ready for review again?

**Comment by dklimpel** [comment]
> @dklimpel Is this ready for review again?

Yes, and a question: https://github.com/matrix-org/synapse/pull/9045#discussion_r555184753

File: tests/rest/admin/test_user.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (clokep) [reply tests/rest/admin/test_user.py:747]: I don't have a strong opinion about it, I think the way you've done it is fine.

**Review by clokep — APPROVED** [review]
LGTM! Sorry for the delay on this! 👍
